### ISSUE TITLE
Locating KDC's by DNS SRV records

### ISSUE THREAD

**Issue by bodgit** [jcmturner/gokrb5#57]
Once I figured out #56 my next error is that my test client gets the error `client krb5 config does not have any defined KDCs for the default realm`. My `/etc/krb5.conf` simply has the following:
```
[libdefaults]
 dns_lookup_realm = true
 dns_lookup_kdc = true
 default_realm = EXAMPLE.COM
---8<---

[realms]
 EXAMPLE.COM = {
 }
```
which means the KDC's are located using DNS SRV records, i.e. `_kerberos._tcp.example.com`, (it's a Windows AD environment).

For now I've just done the lookup manually with `dig` and added them to the configuration file but would you be open to this as an enhancement?

**Comment by bodgit** [comment]
I'm happy to have a stab at trying to get this working using github/miekg/dns to do the DNS heavy lifting however the only bit I'm not sure about is the correct way to get from the realm to the DNS domain. Even though I can clearly see in 99% of the cases it just differs by case there are probably some heuristics involved.

**Comment by jcmturner** [comment]
I'm soon to commit something on this so you probably want to hold off. I'll put something up on a branch but I want to create some integration tests before I merge to master.

**Comment by bodgit** [comment]
> I'm soon to commit something on this so you probably want to hold off. I'll put something up on a branch but I want to create some integration tests before I merge to master.

Ok, happy to give it a test.

> Even though I can clearly see in 99% of the cases it just differs by case there are probably some heuristics involved.

After a bit more searching, https://github.com/krb5/krb5/blob/a590ad732ece9da7831673fa472769b6fb319934/doc/admin/realm_config.rst#hostnames-for-kdcs states it's as simple as looking up `_kerberos._udp.REALM`, iff the realm is a DNS domain name.

**Comment by jcmturner** [comment]
@bodgit There is something on branch issue-57 that I think should work. I've not completed all the testing yet. It only works for Go version 1.9 atm. I need to think about how to integration test this. If you feel like trying it and letting me know if it works for you that'd be useful.

**Comment by bodgit** [comment]
A quick cursory test with all of the `kdc = ...` entries commented out in `/etc/krb5.conf` seems to still work. If I run `tcpdump -nn -i any -s 0 -X port 53` alongside my test code I can see the various SRV lookups being made followed by AAAA and A lookups when it picks one of the returned KDC hostnames.

**Comment by jcmturner** [comment]
OK sounds good. The challenge I have is in how to easily test the DNS integration.
I wrote this: https://github.com/jcmturner/dnsutils/blob/master/srv.go
to get the SRV records and order them according to priority and weight. To easily integration test I found a way to override the OS level name server configuration to test against a bind docker container. However this way of overriding is only available in Go 1.9 so I'm not sure how I write this to be automatically testable in 1.8 and 1.7...
